exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/RowKeyJoinPrel.java

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.planner.physical;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+import org.apache.drill.exec.physical.base.PhysicalOperator;
+import org.apache.drill.exec.physical.config.RowKeyJoinPOP;
+import org.apache.drill.exec.planner.cost.DrillCostBase.DrillCostFactory;
+import org.apache.drill.exec.planner.physical.visitor.PrelVisitor;
+import org.apache.drill.exec.record.BatchSchema.SelectionVectorMode;
+import org.apache.calcite.rel.InvalidRelException;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelTraitSet;
+
+import org.apache.drill.shaded.guava.com.google.common.base.Preconditions;
+
+public class RowKeyJoinPrel extends JoinPrel implements Prel {
+
+  double estimatedRowCount = -1;
+  public RowKeyJoinPrel(RelOptCluster cluster, RelTraitSet traits, RelNode left, RelNode right,
+      RexNode condition, JoinRelType joinType) throws InvalidRelException {
+    super(cluster, traits, left, right, condition, joinType);
+    Preconditions.checkArgument(joinType == JoinRelType.INNER);
+  }
+
+  @Override
+  public PhysicalOperator getPhysicalOperator(PhysicalPlanCreator creator) throws IOException {
+    PhysicalOperator leftPop = ((Prel)left).getPhysicalOperator(creator);
+    PhysicalOperator rightPop = ((Prel)right).getPhysicalOperator(creator);
+    RowKeyJoinPOP rkPop = new RowKeyJoinPOP(leftPop, rightPop);
+    return creator.addMetadata(this, rkPop);
+  }
+
+  @Override
+  public double estimateRowCount(RelMetadataQuery mq) {
+    if (estimatedRowCount >= 0) {
+      return estimatedRowCount;
+    }
+    return this.getLeft().getRows();
+  }
+
+  @Override
+  public Join copy(RelTraitSet traitSet, RexNode conditionExpr, RelNode left, RelNode right,
+      JoinRelType joinType, boolean semiJoinDone) {
+    try {
+      RowKeyJoinPrel rkj = new RowKeyJoinPrel(this.getCluster(), traitSet, left, right, conditionExpr, joinType);
+      rkj.setEstimatedRowCount(this.estimatedRowCount);
+      return rkj;
+    } catch (InvalidRelException e) {
+      throw new AssertionError(e);
+    }
+  }
+
+  @Override
+  public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    if(PrelUtil.getSettings(getCluster()).useDefaultCosting()) {
+      return super.computeSelfCost(planner).multiplyBy(.1);
+    }
+    double rowCount = mq.getRowCount(this.getRight());
+    DrillCostFactory costFactory = (DrillCostFactory) planner.getCostFactory();
+    return costFactory.makeCost(rowCount, 0, 0, 0,
```

Comment:
RowKeyJoin operator by itself incurs negligible CPU and I/O cost since it is not doing a real join.  The actual cost is attributed to the skip-scan (random I/O).  The RK join will hold 1 batch in memory but since it is not a buffering operator, we typically don't assign memory costs to them.   I can add the explanation.